### PR TITLE
EECO-473 Add support for SNS URI validation

### PR DESCRIPTION
**Pull request by capripot**

### CODE REVIEW

File: lib/aws/sns/uri.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
+require "aws-sdk-sns"
+require "addressable/uri"
+
+module Aws
```
Comment (admytren) [inline lib/aws/sns/uri.rb:4]: Should this be in "Aws::SNS" namespace? I got a bit confused earlier and was looking for "URI" class in Amazon docs: https://docs.aws.amazon.com/sdkforruby/api/Aws/SNS.html

File: lib/aws/sns/uri.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        if uri.user.blank? || uri.password.blank?
+          raise Aws::SNS::Errors::InvalidURIError, "User or Password missing"
+        end
+
+        key_id = CGI.unescape(uri.user)
+        access_key = CGI.unescape(uri.password)
+
+        segments = uri.path&.split('/')
+
+        raise Aws::SNS::Errors::InvalidURIError unless segments&.size == 3
```
Comment (admytren) [inline lib/aws/sns/uri.rb:27]: Would it be good to add message saying exactly 3 segments were expected in URI?

File: lib/aws/sns/uri.rb [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+        topic_owner = CGI.unescape(segments[1])
+        topic_name = CGI.unescape(segments[2])
+
+        new(
+          key_id: key_id,
+          access_key: access_key,
+          region: region,
```
Comment (admytren) [inline lib/aws/sns/uri.rb:35]: Is it okay for region to be empty?

File: lib/aws/sns/uri.rb [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        @region = region
+        @topic_owner = topic_owner
+        @topic_name = topic_name
+      end
+
+      def arn
+        "arn:aws:sns:#{region}:#{topic_owner}:#{topic_name}"
+      end
+
+      def test_publish!
```
Comment (admytren) [inline lib/aws/sns/uri.rb]: Ruby newbie question: is the reason why you're adding "!" to the method name to hint that this method would mutate external state (e.g. send SNS to uri being tested?)

File: spec/validate_url_spec.rb [hunk 5]
```diff
@@ -271,4 +271,20 @@
       expect(@user.errors[:homepage]).to eq(["wrong"])
     end
   end
+
+  context "with SNS" do
+    before do
+      @user = UserWithArAndSns.new
+    end
+
+    it "should not allow a valid sns uri" do
```
Comment (admytren) [inline spec/validate_url_spec.rb]: Recommend adding more context in the test description here and on line 285 (E.g. which part of uri is broken)

**Review by admytren — APPROVED** [review]

File: lib/aws/sns/uri.rb (hunk 2 above)
Comment (capripot) [reply lib/aws/sns/uri.rb:27]: I think the format error default message makes sense in that case. What do you think?

File: lib/aws/sns/uri.rb (hunk 4 above)
Comment (capripot) [reply lib/aws/sns/uri.rb]: The original meaning is mutate the object it's called on actually. Maybe I should not use it here actually. Good point!

File: lib/aws/sns/uri.rb (hunk 1 above)
Comment (capripot) [reply lib/aws/sns/uri.rb:4]: I think it's fine, it's sort of an amendment to the gem. `show-source` in pry helps when it comes to find where it's defined at runtime.

File: lib/aws/sns/uri.rb (hunk 3 above)
Comment (capripot) [reply lib/aws/sns/uri.rb:35]: Fixed it

File: lib/aws/sns/uri.rb (hunk 2 above)
Comment (admytren) [reply lib/aws/sns/uri.rb:27]: Yeah that's true, it will explain there what's expected in the SNS URI

**Pull request merged**